File: frontend/src/types/signal.ts
export type SignalDirection = "buy" | "sell" | "neutral";
export type SignalTradeStyle = "scalp" | "swing";
export type SignalStatus = "active" | "watchlist" | "expired";
export type Timeframe = "1m" | "5m" | "15m" | "30m" | "1h" | "4h" | "1d";

/**
 * What price did after the signal was generated — the backend outcome
 * vocabulary (see `models/signal.py::SignalOutcome`). `open` until the outcome
 * evaluator finds a terminal result.
 */
export type SignalOutcome =
  | "open"
  | "hit_tp1"
  | "hit_tp2"
  | "hit_tp3"
  | "hit_sl"
  | "expired"
  | "cancelled";

export type SignalTargetLabel = "TP1" | "TP2" | "TP3";

export type SignalTarget = {
  label: SignalTargetLabel;
  price: number;
  /** Signed distance from entry, as a fraction (e.g. 0.012 = +1.2%). */
  distancePercent: number | null;
};

export type SignalReasoning = {
  thesis: string;
  confirmations: string[];
  riskPlan: string;
  invalidation: string;
  executionNotes: string[];
};

/**
 * Technical indicators captured at signal-generation time. A normalised,
 * UI-facing mirror of the backend `indicators_snapshot` JSONB payload.
 */
export type IndicatorSnapshot = {
  asOf: string | null;
  candlesAnalyzed: number;
  lastClose: number | null;
  sma20: number | null;
  sma50: number | null;
  ema20: number | null;
  ema50: number | null;
  ema200: number | null;
  rsi14: number | null;
  macd: number | null;
  macdSignal: number | null;
  macdHistogram: number | null;
  atr14: number | null;
  bbUpper: number | null;
  bbMiddle: number | null;
  bbLower: number | null;
  bbPercent: number | null;
};

export type TradingPair = {
  id: number;
  symbol: string;
  baseCurrency: string;
  quoteCurrency: string;
  displayName: string;
  isActive: boolean;
};

export type Signal = {
  id: string;
  pairId: number;
  analysisRunId: string | null;
  symbol: string;
  displayName: string;
  direction: SignalDirection;
  tradeStyle: SignalTradeStyle;
  status: SignalStatus;
  confidence: number;
  entryPrice: number;
  stopLoss: number | null;
  /** Signed distance from entry to stop, as a fraction. Null without a stop. */
  stopDistancePercent: number | null;
  targets: SignalTarget[];
  timeframe: Timeframe;
  generatedAt: string;
  expiresAt: string | null;
  riskReward: number | null;
  rationale: string;
  reasoning: SignalReasoning;
  indicators: IndicatorSnapshot | null;
  aiProvider: string | null;
  aiModel: string | null;
  /** What happened after generation. `open` until the evaluator closes it. */
  outcome: SignalOutcome;
  /** Realised result in R multiples. Null while open or when risk is undefined. */
  realizedR: number | null;
  /** When the signal reached a terminal outcome (ISO 8601). Null while open. */
  closedAt: string | null;
};

export type SignalStats = {
  activeSignals: number;
  averageConfidence: number;
  modelWinRate: number;
  monitoredPairs: number;
  analysisCadenceMinutes: number;
  monthlySignalVolume: number;
};

File: backend/app/schemas/signal.py
"""Wire-format models for signals — the shape the API surfaces to the frontend.

This module is transport-agnostic and deliberately decoupled from persistence:
per the layering table in ``backend/README.md``, schemas may import ``pydantic``
and stdlib only — never ``sqlalchemy``, the ORM models, or the service layer.
That is why ``direction`` is a plain ``Literal`` here rather than a reuse of the
ORM ``SignalDirection`` enum, and why the ORM→wire mapping lives in the
controller, not in a ``from_attributes`` hook on this model: keeping the
translation explicit (and in the layer allowed to see both sides) is what stops
a stray ``signals`` column change from silently reshaping the public contract.

Money fields are ``Decimal``. They serialise to JSON **strings**, not numbers —
the same "never float for prices" discipline the storage layer keeps (see the
``Numeric`` columns on the ``Signal`` model). Clients parse them to a number at
the display boundary, mirroring how the backend only drops to ``float`` at the
pandas boundary in the indicator calculator.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field

SignalDirection = Literal["buy", "sell", "neutral"]
SignalType = Literal["scalp", "swing"]
SignalOutcome = Literal[
    "open",
    "hit_tp1",
    "hit_tp2",
    "hit_tp3",
    "hit_sl",
    "expired",
    "cancelled",
]

# Coarse, lifecycle-derived status the browse UI filters by (computed from
# direction + expiry, not stored): a directional, unexpired signal is "active";
# a neutral one is "watchlist"; a lapsed one is "expired".
SignalStatusFilter = Literal["active", "watchlist", "expired"]
# Outcome grouped into result categories for filtering (a "win" is any
# take-profit rung, a "loss" is the stop, "expired" covers expired/cancelled).
SignalResultFilter = Literal["open", "win", "loss", "expired"]
# Server-side ordering for the signal list.
SignalSort = Literal["confidence", "newest", "symbol"]


class SignalResponse(BaseModel):
    """A persisted trade signal as returned by the read endpoints.

    ``pair_symbol`` is denormalised onto the response so the frontend can render
    a signal without a second round trip to resolve the pair — the controller
    fills it from the eagerly-loaded ``Pair`` while the session is still open.
    """

    id: uuid.UUID
    pair_id: int
    pair_symbol: str | None = None
    analysis_run_id: uuid.UUID | None = None

    direction: SignalDirection
    # Trade horizon this signal is framed on. A run emits one of each style per
    # pair (scalp = short-term/lower TF, swing = higher TF).
    signal_type: SignalType
    confidence: float = Field(ge=0.0, le=1.0)

    entry_price: Decimal
    stop_loss: Decimal | None = None
    # The take-profit ladder, ordered TP1..TP3. ``take_profit`` is the primary
    # target (TP1); the secondary scale-out targets are nullable because a
    # signal may carry fewer than three.
    take_profit: Decimal | None = None
    take_profit_2: Decimal | None = None
    take_profit_3: Decimal | None = None

    timeframe: str
    rationale: str | None = None
    # The raw indicator values that fed the AI prompt at generation time —
    # surfaced verbatim so the frontend can explain *why* a signal exists.
    indicators_snapshot: dict[str, Any] | None = None

    generated_at: datetime
    expires_at: datetime | None = None

    ai_provider: str | None = None
    ai_model: str | None = None

    # ── Outcome tracking (Iteration 7) ────────────────────────────────────
    # What price did after the signal was generated. ``outcome`` is ``open``
    # until the evaluator finds a terminal result; ``realized_r`` (in R
    # multiples) and ``closed_at`` are populated only once it closes.
    outcome: SignalOutcome = "open"
    realized_r: Decimal | None = None
    closed_at: datetime | None = None
